### ISSUE TITLE
Clarification for a case of open recursion 

### ISSUE THREAD

**Issue by pavelrappo** [reactive-streams/reactive-streams-jvm#429]
Consider a `Publisher`-`Subscriber` couple whose goal is to process a finite ordered sequence of elements. If I understand correctly, there might be otherwise perfect (i.e. TCK-compliant) implementations of `Publisher` which emit elements via open recursion: 
```Subscriber.onNext
    Subscription.request
        Subscriber.onNext
            Subscription.request
                ...
```
Now the `Subscriber` might implement `onNext` such that it first replenishes the demand (synchronously) and then processes the element: 
```java
            @Override
            public void onNext(T element) {
                subscription.request(1);
                process(element);
            }
```
or the other way around: 
```java
            @Override
            public void onNext(T element) {
                process(element);
                subscription.request(1);
            }
```
Unless additional actions are taken, the results might be quite different. 
The question is this. Whose burden is it to make sure the order in that sequence does not depend on the choice of the processing style (assuming that preserving this order is significant)?

**Comment by akarnokd** [comment]
The TCK defaults to limited recursion so your `Processor` wouldn't pass it. See rule [§3.2](https://github.com/reactive-streams/reactive-streams-jvm#3.2).

> Whose burden is it to make sure the order in that sequence does not depend on the choice of the processing style (assuming that preserving this order is significant)?

It's always the responsibility of the implementor of the `Publisher`/`Processor`.

**Comment by pavelrappo** [comment]
> The TCK defaults to limited recursion so your Processor wouldn't pass it
I guess, it all depends on a semantics of the concrete `Processor`. My toy example `Processor` passes TCK (modulo, Subscribe rules, which are not important to this conversation).
```java
public class TenElementsSequencePublisher<T> implements Publisher<T> {

    private final List<T> list;

    public TenElementsSequencePublisher(List<T> list) {
        this.list = Objects.requireNonNull(list);
    }

    @Override
    public void subscribe(Subscriber<? super T> subscriber) {
        subscriber.onSubscribe(new Subscription() {

            private final Iterator<T> iterator = list.iterator();
            volatile boolean cancelled;

            @Override
            public void request(long n) {
                if (cancelled) {
                    return;
                }
                if (n <= 0) {
                    cancel();
                    subscriber.onError(new IllegalArgumentException(
                            "non-positive subscription request"));
                    return;
                }
                long d = n;
                while (!cancelled) {
                    if (iterator.hasNext()) {
                        if (d > 0) {
                            d--;
                            subscriber.onNext(iterator.next());
                            continue;
                        }
                    } else {
                        cancel();
                        subscriber.onComplete();
                    }
                    break;
                }
            }

            @Override
            public void cancel() {
                cancelled = true;
            }
        });
    }
```
```java
public class TenElementsSequencePublisherVerification extends PublisherVerification<Long> {

    public TenElementsSequencePublisherVerification() {
        super(new TestEnvironment());
    }

    @Override
    public long maxElementsFromPublisher() {
        return 10;
    }

    @Override
    public long boundedDepthOfOnNextAndRequestRecursion() {
        return 10;
    }

    @Override
    public Publisher<Long> createPublisher(long elements) {
        return new TenElementsSequencePublisher<>(
                LongStream.range(0, elements).boxed().collect(Collectors.toList()));
    }

    @Override
    public Publisher<Long> createFailedPublisher() {
        return null;
    }
}
```

**Comment by akarnokd** [comment]
You implementation is not thread safe, any asynchronous request coming when the emission is ongoing will break it. You can find a safe version in RxJava ([thread safety](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/flowable/FlowableFromIterable.java#L121), [emit based on request count](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/flowable/FlowableFromIterable.java#L121)).

Also increasing the TCK recursion limit will risk a `StackOverflowError` with long flows.

**Comment by pavelrappo** [comment]
David, it is not essential for this topic I believe. What asynchronous request are you talking about? Could you please provide an example?

**Comment by akarnokd** [comment]
```java
@Override
public void onNext(T element) {
    executor.submit(() -> {
        subscription.request(1);
        process(element);
    });
}
```

**Comment by pavelrappo** [comment]
Hm... I thought §2.7 prohibits such tricks:

> A Subscriber MUST ensure that all calls on its Subscription take place from the same thread or provide for respective external synchronization.

So in your example the `Subscriber` does not provide any external synchronization.

**Comment by akarnokd** [comment]
The `executor` does if it is a single-threaded one. If it weren't, existing reactive libraries defend against async request calls with trampolining (RxJava, Reactor) or message queue of the backing actor (Akka Stream). The thing is, you can't know and "external synchronization" can simply mean the `Subscription` can take steps to ensure thread safety.

Btw, why don't you use any existing Reactive Streams compliant libraries where far more complicated race conditions were already worked out?

**Comment by pavelrappo** [comment]
But David, if it is a single-threaded executor (if I read you right), then this `Subscriber` seems to be effectively thread-safe. What am I missing?

> Btw, why don't you use any existing Reactive Streams compliant libraries where far more complicated race conditions have been already worked out?

The reason is that I was re-reading the spec because I spotted this ordering thing. And asked a question. I don't need an RS solution right now. But if I do, I will certainly use what you are saying.

**Comment by akarnokd** [comment]
The `Subscriber` is thread safe, but your `Subscription` isn't. The `onNext` call goes to another thread and that thread executes `request(1)` while the original thread just returns from your `subscriber.onNext`. Now if the initial request was 1, both threads can execute `iterator.hasNext()` concurrently. If the initial request was more than one, you can get `iterator.hasNext()` executing concurrently with `iterator.next()`.

**Comment by pavelrappo** [comment]
But hey, then there's no synchronization in between these threads. `T1` who initially called `onNext` and `T2` this executor's worker thread. We are back to square one and the rule §2.7. I will wait for the spec authors to explain the value of this rule. I don't seem to understand it.

**Comment by jroper** [comment]
Let me check that I understand this... you're saying that you expect these two implementations:

```java
            @Override
            public void onNext(T element) {
                process(element);
                subscription.request(1);
            }
```

and

```java
            @Override
            public void onNext(T element) {
                subscription.request(1);
                process(element);
            }
```

to have the same behaviour? The spec explicitly states that they don't, see §3.10:

bq. While the `Subscription` is not cancelled, `Subscription.request(long n)` MAY synchronously call `onNext` on this (or other) subscriber(s).

So, that is to say, when a `Subscriber` invokes `Subscription.request`, that subscriber must be prepared to synchronously (ie recursively) receive another `onNext` signal. So, if there is any processing that the subscriber needs to do before it receives the next `onNext` signal, the subscriber has to ensure that it has already done that before it invokes `request`.

I believe @akarnokd's point is that for most RS implementations, this is just not an issue because most RS implementations dispatch the operation to something that ensures sequential invocation (eg a trampolining mutex executor or a message queue), since they need to do that anyway to implement the thread safety guarantees required by the spec (ie, happens before, exclusion of mutual exclusion of method invocations, etc) and the infinite recursion prevention anyway.

**Comment by pavelrappo** [comment]
James, I must admit it indeed follows from the specification that a `Subscriber` must be prepared for possible differences in implementations of `Publisher`s. However, I believe it deserves a word of caution (at the very least). I would expect the specification to explicitly assign the responsibility to either a `Publisher` or a `Subscriber`. 

Look at it like this. Even you and David answered differently. As David said:
> It's always the responsibility of the implementor of the Publisher/Processor.

From your answer (and the spec) it follows that it's the responsibility of a `Subscriber`. To me it alone is alarming. The fact that the real life `Publisher`s provide this guarantee anyway (probably as a by-product of their sophisticated implementations) cannot be satisfactory. 

I'm aware you see Reactive Streams as an integration API rather than an end-user-facing one. If that is true, then it shouldn't really matter which participant is harder to implement -- a `Publisher` or a `Subscriber` as the user will do neither. If a `Publisher` is allowed to be so primitive as to simply recurse, it means the `Subscriber` has to be defensive. The examples of a `Subscriber` I've seen so far do not do anything special in this regard. It either means they do not care of misordered deliveries or do not know they are faulty and waiting for a recursive publisher to show them that.

**Comment by akarnokd** [comment]
I'm confident `Subscriber`s can't do much about it because to avoid recursion, they would need somehow to make sure `request()` happens outside of the execution of their `onNext` method so that the call doesn't trigger re-entrance. If you call `request` from within `onNext`, the `Publisher` has to know it is already fulfilling a former request right then and not call `onNext` recursively, which would violate the happens-before requirement of the spec.

A solution could have been to return a non-zero number to have the `Publisher` send that many more, but such approach forces synchronous communication between the `Publisher` and `Subscriber` as well as prevents any form delayed requesting [like this](https://github.com/reactive-streams/reactive-streams-jvm/issues/429#issuecomment-380565786). (This was one of the main faults of the [OSGi Asynchronous Event Streams](http://akarnokd.blogspot.hu/2015/11/asynchronous-event-streams-vs-reactive.html) proposal).

I've probably said this before, but the Reactive Streams Specification is not constructive: it doesn't give you many tools to implement your own components and can't really guide you to a proper implementation beyond seeing if you pass the TCK or [give some hints about](http://www.reactive-streams.org/reactive-streams-tck-1.0.2-javadoc/org/reactivestreams/tck/PublisherVerification.html#required_spec101_subscriptionRequestMustResultInTheCorrectNumberOfProducedElements--) why you likely didn't pass.

**Comment by pavelrappo** [comment]
> I'm confident Subscribers can't do much about it because to avoid recursion...

There is no need to avoid recusion per se. There is, however, a desire to make sure the order of processing, if significant, is independent from the way the mechanics of the `Publisher` is implemented. I think making sure the processing has been started before replenishing the demand will suffice as a general advice for implementers. Unless, of course, I missed some case (kindly point me to it).

The responsibility of ordering could have been imposed on the `Publisher`, but as far as I can see it now, the authors of the specification decided to go the opposite way and impose it on the `Subscriber`. Maybe to allow the simplest possible implementations of `Publisher`s. I don't know. I don't want to speculate.

The point is, if there's a coupling that integrates a source of data with its target destination I don't want this coupling to interfere with the data. Compare this:
```java        List<Integer> collected = 
                IntStream.range(0, 1024)
                        .boxed()
                        .collect(Collectors.toList());
```
with this
```java
        List<Integer> collected =
                Publishers.fromIterable(() -> IntStream.range(0, 1024).boxed().iterator())
                          .subscribe(Subscribers.toList());
```
There is a great deal of interaction between the `Stream` and the `Collector` and it is well-specified:

> If the stream is parallel, and the Collector is concurrent, and either the stream is unordered or the collector is unordered, then a concurrent reduction will be performed (see Collector for details on concurrent reduction.)

and

> Returns a sequential ordered IntStream from startInclusive (inclusive) to endExclusive (exclusive) by an incremental step of 1.

**Comment by pavelrappo** [comment]
David, could you elaborate on this?

> If you call request from within onNext, the Publisher has to know it is already fulfilling a former request right then and not call onNext recursively, which would violate the happens-before requirement of the spec.

**Comment by rkuhn** [comment]
> The responsibility of ordering could have been imposed on the Publisher, but as far as I can see, the authors of the specification decided to go the opposite way and impose it on the Subscriber. Maybe to allow very simple implementations of Publishers. I don't know. I don't want to speculate.

There is no need to speculate: the first sentence on https://www.reactive-streams.org/ reads “Reactive Streams is an initiative to provide a standard for asynchronous stream processing with non-blocking back pressure.” Elements are passed to an asynchronous subscriber via `onNext`, where the sequence of the `onNext` calls denotes the ordering of events transferred from publisher to subscriber.

In this light, the ordering of events is defined by the Publisher, they are transferred in the publisher’s order over the boundary to the subscriber. It is the subscriber’s responsibility to treat them in that order.

The fact that the spec allows synchronous fulfilment of a request for more elements implies that the subscriber needs to perform the request whenever that is appropriate for correctly ordered processing of the received elements.

These arguments have been had at length in 2014, see for example https://github.com/reactive-streams/reactive-streams-jvm/issues/53. The gist is that RS’s primary motivation is the transport of streams across asynchronous boundaries, where reentrancy simply is not a concern. This is one of the many reasons RS is not an API but rather an SPI for library interoperability—end users cannot routinely be expected to write correct implementations of the RS interfaces.

**Comment by viktorklang** [comment]
Apologies for my late response here, somehow this thread ended up in the wrong folder.

The only thing I have to add to what has already been said in this thread is that the choice of implementing a synchronous Subscriber carries the weight of understanding the implications thereof.

**Comment by viktorklang** [comment]
@pavelrappo Ping

**Comment by pavelrappo** [comment]
@viktorklang, I believe it would be the right thing to do. `Subscriber`'s implementers need to be made aware of possible pitfalls.

**Comment by viktorklang** [comment]
@pavelrappo OTOH perhaps it is better to add it in in the explanation for 2:1, as I imagine that implementors of Subscriber will look at that section first. Thoughts?

**Comment by viktorklang** [comment]
@reactive-streams/contributors @pavelrappo 

What do you think about the following added to the explanation of 2:1

```To avoid reentrancy problems, it is recommended for synchronous Subscriber implementations to invoke `request` at the end of the signal processing method(s).```

**Comment by pavelrappo** [comment]
@viktorklang 
I think it is too vague. We should state exactly what kind of problem we are trying to avoid. Maybe even with an example. I will think of with some wording ideas.

**Comment by viktorklang** [comment]
@pavelrappo I want to avoid having the spec rules turn too wordy, is there something we could refer to for those who might not yet be familiar with reentrancy?

**Comment by pavelrappo** [comment]
@viktorklang, maybe 

`To avoid unwanted signal reordering caused by `Subscription` methods being reentrant, it is RECOMMENDED for `Subscriber` implementations to invoke `Subscription`'s methods at the end of the signal processing.
`
I believe it also affects `cancel`, not just `request`. Some implementations may invoke `onComplete`/`onError` or any other pending signal(s) upon cancellation.

**Comment by viktorklang** [comment]
@pavelrappo Great suggestion! And since the problem only manifests for synchronous Subscriber implementations perhaps something like:

`To avoid unwanted signal reordering caused by reentrant Subscription methods, it is strongly RECOMMENDED for synchronous Subscriber implementations to invoke Subscription methods at the very end of any signal processing.`

**Comment by pavelrappo** [comment]
Thanks, @viktorklang. I believe with your changes the wording is good now.

**Comment by viktorklang** [comment]
@pavelrappo Perfect, thank you!

@reactive-streams/contributors What do you say w.r.t. this proposed change?

**Comment by DougLea** [comment]
I think this sounds OK. If there were a good place to mention it, it would also be worthwhile to tell people not to use request(1) unless they really mean it.  Unnecessarily using request(1) (vs periodically requesting > 1) makes r-s  unnecessarily slower than some alternatives.

**Comment by viktorklang** [comment]
@DougLea For synchronous Subscribers it would seem like Long.MaxValue would be the best default—as it will be running on the Publisher thread with natural "back pressure". So request(1) as a default for async Subscribers really only makes sense for buffer-less such. What's your experience with sandwiching a sync Processor buffer in from of such Subscribers? (request(1)'s)

**Comment by akarnokd** [comment]
http://cs.oswego.edu/pipermail/concurrency-interest/2016-September/015355.html

**Comment by viktorklang** [comment]
Thanks everyone!
